services/pdf-ms/lambdas/addPdfToCase.ts

```diff
@@ -1,138 +1,95 @@
 import config from '../../../config.js';
-import * as dynamoDb from '../../../libs/dynamoDb.js';
-import { Template, FormType, Case, formTypes, AnswerObject } from '../helpers/types';
+import params from '../../../libs/params';
+
+import { Template } from '../helpers/types';
+import { arrayToObject } from '../helpers/caseDataConverter';
 import { loadFileFromBucket, writeFileToBucket } from '../helpers/s3';
 import { modifyPdf } from '../helpers/pdf';
 import { getUser } from '../helpers/user';
-import { getNewAndChangedValues, getApplicantCasesByFormId } from '../helpers/case';
+import {
+  getNewAndChangedCaseAnswerValues,
+  getUserCases,
+  sortCasesByDate,
+  addPdfToCase,
+} from '../helpers/case';
+
+const PDF_SSM_PARAMS = params.read(config.pdf.envsKeyName);
 
 const dynamoDbConverter = AWS.DynamoDB.Converter;
 
-export const main = async (event: Record<string, any>) => {
-  const caseData: Case & { user?: Record<string, any>; answersArray: AnswerObject[] } = parseCase(
+export async function main(event: Record<string, any>): Promise<Boolean> {
+  // Convert DynamoDB case data to plain object
+  const unMarshalledCase: Record<string, any> = dynamoDbConverter.unmarshall(
     event.detail.dynamodb.NewImage
   );
-  if (caseData.pdfGenerated) {
-    return;
+
+  if (unMarshalledCase.pdfGenerated === 'yes') {
+    return true;
   }
-  const { formId, PK, SK } = caseData;
+
+  const { PK } = unMarshalledCase;
   const personalNumber = PK.substring(5);
-  const user = await getUser(personalNumber);
-  if (user) {
-    caseData.answers.user = user;
+
+  const [getUserCasesError, userCases] = await to(getUserCases(personalNumber));
+  if (getUserCasesError) {
+    console.error(getUserCasesError);
+    throw getUserCasesError;
   }
 
-  // find changed and newly added values, compared to the last case with the same formId.
-  const relevantCases = await getApplicantCasesByFormId(personalNumber, formId);
-  let changedValues: string[] = [];
-  let newValues: string[] = [];
-  if (relevantCases.length > 1) {
-    const sortedCases = relevantCases
-      .filter(c => c.status.type.includes('submitted'))
-      .filter(c => c.SK !== SK)
-      .sort((c1, c2) => c2.updatedAt - c1.updatedAt);
-    const newestRelevantCase = sortedCases[0];
-    const newAndChangedValues = getNewAndChangedValues(caseData, newestRelevantCase);
-    changedValues = newAndChangedValues.changedValues;
-    newValues = newAndChangedValues.newValues;
-    caseData.answers.valuesChanged =
-      changedValues.length > 0 || newValues.length > 0 ? 'Ja' : 'Nej';
+  const [currentCase, previousCase] = sortCasesByDate(userCases, 2);
+
+  const ssmParams = await PDF_SSM_PARAMS;
+  const currentCaseAnswers = currentCase.forms[ssmParams.recurringFormId].answers;
+  const previousCaseAnswers = previousCase.forms[ssmParams.recurringFormId].answers;
+
+  const { changedValues, newValues } = getNewAndChangedCaseAnswerValues(
+    currentCaseAnswers,
+    previousCaseAnswers
+  );
```

Comment:
This breaks if the user only has a single case. In that case, the previousCase becomes undefined, and when accessing previousCase.forms[...] it crashes. So, we need to handle that, and decide if in this case we want everything to be colored red or everything colored black.


In general I find the logic here a little suspect. We simply get all cases on the user, but there's no guarantee that the cases will contain the relevant form, so even if the case exists, the previousCase.forms[ssmParams.recurringFormId] might not exist, so there needs to be some filtering on the cases to only pick those that contain the correct formId in forms.